# ASSIGNMENT-1 LINEAR REGRESSION

Dataset: data.csv
Tasks to be performed:
1. Load the dataset using pandas

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score

In [2]:
# Load the dataset using pandas
data = pd.read_csv('data.csv')

In [3]:
# Check the first 5 rows of the data
data.head()

YearsExperience   Salary
0              1.1  39343.0
1              1.3  46205.0
2              1.5  37731.0
3              2.0  43525.0
4              2.2  39891.0

2. Extract data fromYearsExperience column is a variable named X

In [4]:
X = data[['YearsExperience']]

3. Extract data from salary column is a variable named Y

In [5]:
Y = data['Salary']

4. Divide the dataset into two parts for training and testing in 66% and 33% proportion

In [6]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.33, random_state = 42)

5. Create and train LinearRegression Model on training set

In [7]:
model = LinearRegression()
model.fit(X_train, Y_train)

LinearRegression()

6. Make predictions based on the testing set using the trained model

In [8]:
y_pred = model.predict(X_test)

7. Check the performance by calculating the r2 score of the model

In [12]:
r2 = r2_score(Y_test, y_pred)
print("r2 score of the model:", r2)

r2 score of the model: 0.9553063138044949


The R-squared value of the model is 0.9553, which indicates that the model explains 95.53% of the variance in the test data. This means that the model is a good fit for the data and can accurately predict the salaries of employees based on their years of experience.

In conclusion, the linear regression model trained on the given dataset is a reliable method for predicting salaries based on years of experience.

# ASSIGNMENT-2 LOGISTIC REGRESSION

Dataset: diabetes.csv
Tasks to be performed:
1. Load the dataset using pandas 

In [14]:
# Load the dataset using pandas
df = pd.read_csv('diabetes.csv')

# Check the first 5 rows of the dataset
df.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

2. Extract data fromOutcome column is a variable named Y 

In [20]:
Y = df['Outcome']

# Check first 5 rows of Y variable
print(Y.head())

0    1
1    0
2    1
3    0
4    1
Name: Outcome, dtype: int64


3. Extract data from every column except Outcome column in a variable named X 

In [21]:
X = df.drop('Outcome', axis = 1)

# Check first 5 rows of X variable
print(X.head())

   Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  
0                     0.627   50  
1                     0.351   31  
2                     0.672   32  
3                     0.167   21  
4                     2.288   33  


4. Divide the dataset into two parts for training and testing in 70% and 30% proportion 

In [24]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.30, random_state = 42)

5. Create and train Logistic Regression Model on training set 

In [27]:
# Importing the libraries
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix, accuracy_score

model = LogisticRegression()
model.fit(X_train, Y_train)

/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

6. Make predictions based on the testing set using the trained model 

In [28]:
y_pred = model.predict(X_test)

7. Check the performance by calculating the confusion matrix and accuracy score of the model 

In [31]:
cm = confusion_matrix(Y_test, y_pred)
acc_score = accuracy_score(Y_test, y_pred)

In [32]:
print("Confusion Matrix:")
print(cm)
print("Accuracy Score:", acc_score)

Confusion Matrix:
[[121  30]
 [ 30  50]]
Accuracy Score: 0.7402597402597403


Based on the confusion matrix and accuracy score, we can conclude that the logistic regression model has a moderate performance on the diabetes dataset. The confusion matrix shows that out of the 231 test samples, 121 were correctly classified as not having diabetes (true negatives), while 50 were correctly classified as having diabetes (true positives). However, 30 cases were incorrectly classified as having diabetes (false positives), while 30 cases were incorrectly classified as not having diabetes (false negatives).

The accuracy score of the model is 0.74, which indicates that it correctly classified 74% of the test samples. While this accuracy is moderate, it may not be sufficient for critical medical diagnoses. In such cases, we may need to consider other evaluation metrics such as precision, recall, and F1 score to evaluate the performance of the model.

# ASSIGNMENT-3 DECISION TREE

Dataset: diabetes.csv
Tasks to be performed:
1. Load the dataset using pandas

In [35]:
#Importing the libraries
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import confusion_matrix, accuracy_score

# lOADING THE DATASET USING PANDAS
df = pd.read_csv('diabetes.csv')

# Check first 5 rows of the dataset
df.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

2. Extract data fromOutcome column is a variable named Y 

In [37]:
Y = df['Outcome']

# Check first 5 rows of Y variable
print(Y.head())

0    1
1    0
2    1
3    0
4    1
Name: Outcome, dtype: int64


3. Extract data from every column except Outcome column in a variable named X

In [39]:
X = df.drop('Outcome', axis = 1)

# Check first 5 rows of Y variable
print(X.head())

   Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  
0                     0.627   50  
1                     0.351   31  
2                     0.672   32  
3                     0.167   21  
4                     2.288   33  


4. Divide the dataset into two parts for training and testing in 70% and 30% proportion 

In [40]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.30, random_state = 42)

5. Create and train Decision Tree Model on training set 

In [41]:
model = DecisionTreeClassifier()
model.fit(X_train, Y_train)

DecisionTreeClassifier()

6. Make predictions based on the testing set using the trained model 

In [42]:
Y_pred = model.predict(X_test)

7. Check the performance by calculating the confusion matrix and accuracy score of the model 

In [43]:
cm = confusion_matrix(Y_test, Y_pred)
accu_score = accuracy_score(Y_test, Y_pred)

In [45]:
print('Confusion_matrix: ')
print(cm)
print('Accoracy Score = ', accu_score)

Confusion_matrix: 
[[105  46]
 [ 26  54]]
Accoracy Score =  0.6883116883116883


The confusion matrix shows the number of true positives, false positives, true negatives, and false negatives of the model's predictions. The accuracy score is the ratio of the number of correct predictions to the total number of predictions.

In this case, the decision tree model has an accuracy score of 0.69, which means that it correctly predicted 69% of the outcomes in the testing set. The confusion matrix shows that the model correctly predicted 105 true positives and 54 true negatives, but it incorrectly predicted 46 false positives and 26 false negatives.

Overall, the model's performance is not very good, as it incorrectly predicted a relatively large number of outcomes. It may be necessary to try different models or adjust the hyperparameters of the decision tree model to improve its performance.

# ASSIGNMENT-4 RANDOM FOREST

Dataset: diabetes.csv
Tasks to be performed:
1. Load the dataset using pandas

In [47]:
# Import the libraries
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import confusion_matrix, accuracy_score

# Load the dataset using pandas
df = pd.read_csv('diabetes.csv')

# Check the first 5 rows of the dataset
df.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

2. Extract data fromOutcome column is a variable named Y 

In [48]:
Y = df['Outcome']

# Check first 5 rows of Y variable
print(Y.head())

0    1
1    0
2    1
3    0
4    1
Name: Outcome, dtype: int64


3. Extract data from every column except Outcome column in a variable named X.

In [53]:
X = df.drop('Outcome', axis = 1)

# Check the columns of X variable
print(X.columns)

Index(['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin',
       'BMI', 'DiabetesPedigreeFunction', 'Age'],
      dtype='object')


4. Divide the dataset into two parts for training and testing in 70% and 30% proportion 

In [55]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.30, random_state = 42)

5. Create and train Random Forest Model on training set 

In [59]:
model = RandomForestClassifier(random_state=42)
model.fit(X_train, Y_train)

RandomForestClassifier(random_state=42)

6. Make predictions based on the testing set using the trained model 

In [61]:
Y_pred = model.predict(X_test)

7. Check the performance by calculating the confusion matrix and accuracy score of the model 

In [63]:
cm = confusion_matrix(Y_test, Y_pred)
accu_score = accuracy_score(Y_test, Y_pred)

In [65]:
print("Confusion matrix:\n", cm)
print("Accuracy Score:", accu_score)

Confusion matrix:
 [[121  30]
 [ 27  53]]
Accuracy Score: 0.7532467532467533


Based on the confusion matrix and accuracy score obtained, we can conclude that the Random Forest model performed relatively well in predicting the diabetes outcome. The model correctly identified 121 instances of non-diabetic patients and 53 instances of diabetic patients, but misclassified 30 instances of non-diabetic patients as diabetic and 27 instances of diabetic patients as non-diabetic. The overall accuracy of the model is 75.32%, which means that the model correctly classified 75.32% of the total instances in the test set.

# ASSIGNMENT-5 Naïve-Bayes Assignment

Dataset: diabetes.csv
Tasks to be performed:
1. Load the dataset using pandas 

In [66]:
# Import the libraries
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import confusion_matrix, accuracy_score

# Loading the dataset using pandas
df = pd.read_csv('diabetes.csv')

# Check first 5 rows of the dataset
df.head()


Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

2. Extract data fromOutcome column is a variable named Y 

In [69]:
Y = df['Outcome']

# Check first 5 rows of Y variable
print(Y.head())

0    1
1    0
2    1
3    0
4    1
Name: Outcome, dtype: int64


3. Extract data from every column except Outcome column in a variable named X

In [70]:
X = df.drop('Outcome', axis = 1)
print(X.columns)

Index(['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin',
       'BMI', 'DiabetesPedigreeFunction', 'Age'],
      dtype='object')


4. Divide the dataset into two parts for training and testing in 70% and 30% proportion 

In [71]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.30, random_state = 42)

5. Create and train Naïve Bayes Model on training set 

In [72]:
model = GaussianNB()
model.fit(X_train, Y_train)

GaussianNB()

6. Make predictions based on the testing set using the trained model 

In [74]:
Y_pred = model.predict(X_test)

7. Check the performance by calculating the confusion matrix and accuracy score of the model 

In [75]:
cm = confusion_matrix(Y_test, Y_pred)
acc_score = accuracy_score(Y_test, Y_pred)

In [77]:
print("Confusion matrix:\n", cm)
print("Accuracy Score:", acc_score)

Confusion matrix:
 [[119  32]
 [ 27  53]]
Accuracy Score: 0.7445887445887446


The Naive Bayes model has an accuracy of 0.7446, which is not as high as the accuracy achieved by the Random Forest model (0.7532) but is still better than the accuracy achieved by the Decision Tree model (0.6883). The model correctly classified 119 out of 151 negative cases and 53 out of 80 positive cases. The confusion matrix shows that the model made 32 false negative predictions and 27 false positive predictions. This indicates that the model has a higher tendency to incorrectly predict positive cases as negative (false negatives) compared to incorrectly predicting negative cases as positive (false positives). Overall, the Naive Bayes model has performed reasonably well in predicting the outcome variable considering its simplicity and quick computation time.

We have evaluated the performance of four machine learning models, i.e., Linear Regression, Logistic Regression, Decision Tree, Random Forest, and Naive Bayes, for predicting the diabetes outcome.

The Linear Regression model explains 95.53% of the variance in the test data, indicating a high level of accuracy in predicting the outcome variable.

The Logistic Regression model achieved an accuracy score of 0.74, indicating moderate performance in predicting diabetes outcomes. The confusion matrix shows that the model correctly classified 121 instances of non-diabetic patients and 50 instances of diabetic patients, but it misclassified 30 instances of non-diabetic patients as diabetic and 30 instances of diabetic patients as non-diabetic.

The Decision Tree model achieved an accuracy score of 0.69, indicating that it was the least accurate model among the four models evaluated. The confusion matrix shows that the model correctly classified 105 true positives and 54 true negatives, but it incorrectly classified 46 false positives and 26 false negatives.

The Random Forest model achieved an accuracy score of 0.7532, which is higher than the Logistic Regression and Decision Tree models. The confusion matrix shows that the model correctly identified 121 instances of non-diabetic patients and 53 instances of diabetic patients, but it misclassified 30 instances of non-diabetic patients as diabetic and 27 instances of diabetic patients as non-diabetic.

The Naive Bayes model achieved an accuracy score of 0.7446, which is also relatively high. The model correctly classified 119 out of 151 negative cases and 53 out of 80 positive cases. However, the model had a higher tendency to incorrectly predict positive cases as negative (false negatives) compared to incorrectly predicting negative cases as positive (false positives).

Overall, among the four models evaluated, the Random Forest model performed the best, followed by the Naive Bayes and Logistic Regression models, while the Decision Tree model performed the worst. However, further analysis and evaluation may be required to determine the best model for the given dataset and task.